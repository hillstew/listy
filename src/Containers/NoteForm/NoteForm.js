import React, { Component } from 'react';
import { Redirect } from 'react-router-dom';
import { connect } from 'react-redux';
import { setError } from '../../actions';
import API from '../../utils/api';
import shortid from 'shortid';

class NoteForm extends Component {
  constructor(props) {
    super(props);
    this.state = {
      showPopUp: true,
      id: this.props.id || -1,
      title: this.props.title || '',
      issues: this.props.issues || [],
    }
  }

  getIndex = (id) => {
    return this.state.issues.findIndex((issue) => issue.id == id);
  }

  createIssuesCopy = () => {
    return this.state.issues.slice();
  }

  handleTitleChange = (e) => {
    this.setState({ title: e.target.value });
  } 

  setIssuesInState = (issues) => {
    this.setState({ issues });
  }

  closePopUp = () => {
    this.setState({ showPopUp: false });
    this.props.fetchNotes();
  }

  handleBodyChange = (e) => {
    const index = this.getIndex(e.target.parentElement.parentElement.id);
    const issues = this.createIssuesCopy();
    issues[index].body = e.target.value;
    this.setIssuesInState(issues);
  } 

  toggleIssueCompletion = (e) => {
    const index = this.getIndex(e.target.parentElement.parentElement.id);
    const issues = this.createIssuesCopy();
    issues[index].completed = !issues[index].completed;
    this.setIssuesInState(issues);
  } 

  handleSubmit = async (e) => {
    e.preventDefault();
    const { setError } = this.props;
    const { id, title, issues } = this.state;
    try {
      id === -1 ?
        await API.fetchData('notes', 'POST', { title, issues }) :
        await API.fetchData(`notes/${id}`, 'PUT', { title, issues });
      this.closePopUp();
    } catch (error) {
      setError(error)
    }
  }

  removeNote = () => {
    if (this.props.id !== -1) {
      this.deleteNote();
    }
    this.closePopUp();
  }

  deleteNote = async () => {
    try {
      await API.fetchData(`notes/${this.state.id}`, 'DELETE');
    } catch (error) {
      setError(error)
    }
  }

  addIssue = (e) => {
    e.preventDefault();
    const issues = this.createIssuesCopy();
    issues.push({ id: shortid.generate(), body: '', completed: false });
    this.setIssuesInState(issues);
  }

  removeIssue = (e) => {
    e.preventDefault();
    const index = this.getIndex(e.target.parentElement.id);
    const issues = this.createIssuesCopy();
    issues.splice(index, 1);
    this.setIssuesInState(issues);
  }

  showIssues = (completed) => {
    const { issues } = this.state;
    return issues.filter(issue => issue.completed === completed).map(issue =>
      <li key={issue.id} id={issue.id}>
        <span>
          <i onClick={this.toggleIssueCompletion} className="fas fa-square"></i>
          <input onChange={this.handleBodyChange} value={issue.body}></input>
        </span>
        <button onClick={this.removeIssue}>Delete</button>
      </li>
    )
  }

  render() {
    const { title, showPopUp } = this.state;
    const incompleteIssues = this.showIssues(false);
    const completeIssues = this.showIssues(true);

    if (!showPopUp) {
      return <Redirect to={'/'} />
    } else {
      return (
        <div className='overlay-div'>
          <form className='note-pop-up' onSubmit={this.handleSubmit}>
            <input onChange={this.handleTitleChange} value={title}></input>
            <ul>{incompleteIssues}</ul>
            <button onClick={this.addIssue}><i className="fas fa-plus-circle form-add-icon"></i></button>
            <h4>Completed</h4>
            <ul>{completeIssues}</ul>
            <input className='submit-button' type="submit" value='Save'></input>
            <button onClick={this.removeNote}><i className="fas fa-trash-alt"></i></button>
          </form>
        </div>
      )
    }
  }
}

const mapDispatchToProps = (dispatch) => ({
  setError: (error) => dispatch(setError(error)),
});

export default connect(null, mapDispatchToProps)(NoteForm);