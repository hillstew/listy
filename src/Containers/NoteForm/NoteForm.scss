.overlay-div {
  position: fixed;
  background-color: rgba(0, 0, 0, 0.5);
  left: 0;
  top: 0;
  width: 100%;
  height: 100%;
}

.note-pop-up {
  position: fixed;
  width: 400px;
  min-height: 400px;
  background-color: white;
  border: 1px solid black;
  border-radius: 10px;
  padding: 20px;
  margin-left: -200px;
  margin-top: -200px;
  left: 50%;
  top: 50%; 
  box-shadow: 2px 2px 8px black;
}

