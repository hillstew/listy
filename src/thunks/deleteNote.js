import { setLoading, setError, removeNote } from '../actions';
import API from '../utils/api';

export const deleteNote = (id) => {
  return async (dispatch) => {
    try {
      dispatch(setLoading(true));
      const result = await API.fetchData(`notes/${id}`, 'DELETE');
      dispatch(setLoading(false));
      dispatch(removeNote(result));
    } catch (error) {
      dispatch(setError(error));
    }
  }
}