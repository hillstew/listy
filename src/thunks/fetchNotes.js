import { setLoading, setError, setNotes } from '../actions'

export const fetchNotes = () => {
  const url = 'http://localhost:3001/api/v1/notes';
  return async (dispatch) =>  {
    try {
      dispatch(setLoading(true))
      const response = await fetch(url)
      if(!response.ok) {
        throw Error(response.statusText)
      }
      dispatch(setLoading(false))
      const notes = await response.json()
      dispatch(setNotes(notes))
    } catch (error) {
      dispatch(setError(error.message))
    }
  }
}