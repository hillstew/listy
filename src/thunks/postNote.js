import { setLoading, setError, addNote, togglePopup } from '../actions';
import API from '../utils/api';

export const postNote = (note) => {
  const { title, issues } = note;
  return async (dispatch) => {
    try {
      dispatch(setLoading(true));
      const result = await API.fetchData('notes', 'POST', { title, issues });
      dispatch(setLoading(false));
      dispatch(addNote(result));
    } catch (error) {
      dispatch(setError(error));
    }
  }
}