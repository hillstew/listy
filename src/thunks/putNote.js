import { setLoading, setError, updateNote } from '../actions';
import API from '../utils/api';

export const putNote = (note) => {
  const { id, title, issues } = note;
  return async (dispatch) => {
    try {
      dispatch(setLoading(true));
      await API.fetchData(`notes/${id}`, 'PUT', { title, issues });
      dispatch(updateNote(note));
    } catch (error) {
      dispatch(setError(error));
    }
    dispatch(setLoading(false));
  }
}