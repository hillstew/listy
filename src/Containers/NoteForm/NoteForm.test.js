import React from 'react'
import { NoteForm, mapStateToProps, mapDispatchToProps } from './NoteForm';
import { shallow } from 'enzyme';
import { putNote } from '../../thunks/putNote';
import { postNote } from '../../thunks/postNote';
import { deleteNote } from '../../thunks/deleteNote';

jest.mock('../../thunks/putNote')
jest.mock('../../thunks/postNote')
jest.mock('../../thunks/deleteNote')

describe('NoteForm', () => {
    describe('NoteForm component', () => {
        let wrapper;

        beforeEach(() => {
            let mockNote = {
                title: 'Title',
                issues: [],
            }
            wrapper = shallow(<NoteForm note={mockNote} />)
        });

        it('should match the correct snapshot', () => {
            expect(wrapper).toMatchSnapshot()
        });
    });

    describe('mapStateToProps', () => {
        it('should return a string with a boolean', () => {
            const mockState = {
                popup: true,
                notes: []
            }
            
            const expected = {
                popup: true,
            }

            const mappedProps = mapStateToProps(mockState);
            expect(mappedProps).toEqual(expected);
        });
    });

    describe('mapDispatchToProps', () => {
        it('should call dispatch with a putNote action when putNote is called', () => {
            const mockDispatch = jest.fn()
            let mockNote = {
                title: 'Title',
                issues: [],
            }
            const actionToDispatch = putNote(mockNote);
            const mappedProps = mapDispatchToProps(mockDispatch);
            mappedProps.putNote(mockNote);
            expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
        });

        it('should call dispatch with a deleteNote action when deleteNote is called', () => {
            const mockDispatch = jest.fn()
            const mockId = 1
            const actionToDispatch = deleteNote(mockId);
            const mappedProps = mapDispatchToProps(mockDispatch);
            mappedProps.deleteNote(mockId);
            expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
        });

        it('should call dispatch with a postNote action when postNote is called', () => {
            const mockDispatch = jest.fn()
            let mockNote = {
                title: 'Title',
                issues: [],
            }
            const actionToDispatch = postNote(mockNote);
            const mappedProps = mapDispatchToProps(mockDispatch);
            mappedProps.postNote(mockNote);
            expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
        });
    });
});