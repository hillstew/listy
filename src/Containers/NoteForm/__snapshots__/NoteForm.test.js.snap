// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NoteForm NoteForm component should match the correct snapshot 1`] = `
<Redirect
  push={false}
  to="/"
/>
`;
